src/model/statement.js

```diff
@@ -0,0 +1,356 @@
+          return new Statement({id, namespace, publisher, timestamp, body, signature})
+        })
+    })
+  }
+
+  calculateSignature (signer: {sign: (msg: Buffer) => Promise<Buffer>}): Promise<Buffer> {
+    const msg: Object = this.toProtobuf()
+    msg.signature = undefined
+    const bytes = pb.stmt.Statement.encode(msg)
+    return signer.sign(bytes)
+  }
+
+  verifySignature (publicKey?: ?PublicSigningKey, keyCache?: Map<string, PublicSigningKey>): Promise<boolean> {
+    return Promise.resolve()
+      .then(() => {
+        keyCache = keyCache || new Map()
+        publicKey = publicKey || keyCache.get(this.publisher) || PublicSigningKey.fromB58String(this.publisher)
+        keyCache.set(this.publisher, publicKey)
+
+        const msg: Object = this.toProtobuf()
+        msg.signature = null
+        const bytes = pb.stmt.Statement.encode(msg)
+        return publicKey.verify(bytes, this.signature)
+      })
+  }
+}
+
+class StatementBody {
+  constructor () {
+    if (this.constructor === StatementBody) {
+      throw new Error("StatementBody class is abstract, don't instantiate directly")
+    }
+  }
+
+  static fromProtobuf (msg: Object): StatementBody {
+    if (msg.simple != null) {
+      return SimpleStatementBody.fromProtobuf((msg.simple: any))
+    } else if (msg.compound != null) {
+      return CompoundStatementBody.fromProtobuf((msg.compound: any))
+    } else if (msg.envelope != null) {
+      return EnvelopeStatementBody.fromProtobuf((msg.envelope: any))
+    }
+
+    throw new Error('Unsupported statement body ' + JSON.stringify(msg))
+  }
+
+  toProtobuf (): StatementBodyMsg {
+    throw new Error('StatementBody.toProtobuf() is not implemented - subclasses must implement')
+  }
+
+  expandObjects (source: Map<string, Object>): StatementBody {
+    return this
+  }
+
+  get refSet (): Set<string> {
+    return new Set()
+  }
+
+  get objectIds (): Array<string> {
+    return []
+  }
+}
+
+class SimpleStatementBody extends StatementBody {
+  objectRef: string
+  refs: Array<string>
+  deps: Array<string>
+  tags: Array<string>
+
+  constructor (contents: {object: string, refs?: Array<string>, deps?: Array<string>, tags?: Array<string>}) {
+    super()
+    this.objectRef = contents.object
+    this.refs = contents.refs || []
+    this.deps = contents.deps || []
+    this.tags = contents.tags || []
+  }
+
+  static fromProtobuf (msg: SimpleStatementMsg): SimpleStatementBody {
+    return new SimpleStatementBody(msg)
+  }
+
+  toProtobuf (): StatementBodyMsg {
+    return {simple: this.toSimpleProtobuf()}
+  }
+
+  toSimpleProtobuf (): SimpleStatementMsg {
+    const {objectRef, refs, deps, tags} = this
+    return {object: objectRef, refs, deps, tags}
+  }
+
+  expandObjects (source: Map<string, Object>): StatementBody {
+    const object = source.get(this.objectRef)
+    if (object == null) {
+      // FIXME: should we just return the un-expanded body instead?
```

Comment:
depends on our treatment of "naked"/"uninflated" objects/statements, I'm inclined to say "yes" but right now we should expect to always have the objects in sane conditions